Added Windows Desktop (WPF) platform implementation

peterfoot opened the pull request:
### Description of Change ###

Adds platform support for Windows desktop (WPF) apps.

### Bugs Fixed ###

- Related to issue #662

Adds the following platform implementations

 - [ ] Accelerometer
 - [x] AppInfo
 - [ ] Barometer
 - [x] Battery
 - [x] Browser
 - [x] Clipboard
 - [ ] Compass
 - [x] Connectivity
 - [x] DeviceDisplay
 - [x] DeviceInfo
 - [x] Email
 - [x] FilePicker
 - [x] FileSystem
 - [ ] Flashlight
 - [ ] Geocoding
 - [x] Geolocation
 - [ ] Gyroscope
 - [x] Launcher
 - [ ] Magnetometer
 - [x] MainThread
 - [x] Map
 - [ ] OrientationSensor
 - [ ] Permissions
 - [x] PhoneDialer
 - [ ] Platform
 - [x] Preferences
 - [x] SecureStorage
 - [x] Screenshot
 - [ ] Share
 - [ ] Shake
 - [x] Sms
 - [x] TextToSpeech
 - [x] VersionTracking
 - [ ] Vibration
 - [ ] WebAuthenticator / AppleSignInAuthenticator

### API Changes ###

Added DevicePlatform.WindowsDesktop

### Behavioral Changes ###

N/A

### PR Checklist ###

- [ ] Has tests (if omitted, state reason in description)
- [x] Has samples (if omitted, state reason in description)
- [x] Rebased on top of master at time of PR
- [x] Changes adhere to coding standard
- [ ] Updated documentation ([see walkthrough](https://github.com/xamarin/Essentials/wiki/Documenting-your-code-with-mdoc))

msftclas commented:
[![CLA assistant check](https://cla.opensource.microsoft.com/pull/badge/not_signed)](https://cla.opensource.microsoft.com/xamarin/Essentials?pullRequest=1079) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.opensource.microsoft.com/xamarin/Essentials?pullRequest=1079) before we can accept your contribution.<br/><br/>:x: peterfoot [sign now](https://cla.opensource.microsoft.com/xamarin/Essentials?pullRequest=1079)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.opensource.microsoft.com/check/xamarin/Essentials?pullRequest=1079) it.</sub>

mattleibow commented on Xamarin.Essentials/Preferences/Preferences.win32.cs:
```diff
@@ -0,0 +1,40 @@
+{
+    public static partial class Preferences
+    {
+        static Dictionary<string, object> preferences = new Dictionary<string, object>();
```
Is this not actually saved to disk at all?

peterfoot replied:
Need some discussion over the implementation of this. This is temporary and required for the code to build and allow the sample app to run. There isn't a native API we can use (unless taking a dependency on the Windows 10 UWP API). A simple implementation would be to persist the dictionary(s) to a file but that may require a dependency on Newtonsoft.Json or similar. I think that will be the easiest option for now and we can discuss further the pros/cons.

Mrnikbobjeff replied:
No need for newtonsoft, BinarySerializer etc. are all available and fit the scenario. Persistance is IMHO necessary for this to be considered workable.

mattleibow replied:
You could have a look here: https://docs.microsoft.com/en-us/dotnet/standard/serialization/

peterfoot replied:
Thanks, I'll rework this and use one of the built-in options.

Mrnikbobjeff replied:
Going the file route comes with another set of problems:
We definitely have to have a lazy mechanism to store and retrieve settings. Loading all properties on first acess would increse the working set unnecessarily. Also, we do need fast key existance checks. Not only for Key enumeration but also for insertion. I am currently hooking up some different benchmarks for a couple of potential solutions, but this all seems very icky to me. Are there any usbale storage technologies available with winapi we could use? or even com perhaps

Mrnikbobjeff replied:
Okay this could be a stupid idea but how about the registry? Fast key existance checks, key value store just as we like it and as persistent as it can be

peterfoot replied:
I think that's frowned upon, we can't control how much it gets used for and unless the app is installed via Centennial it'll get dumped into the main registry.

mattleibow commented on Xamarin.Essentials/Email/Email.win32.cs:
```diff
@@ -0,0 +1,21 @@
+            throw ExceptionUtils.NotSupportedOrImplementedException;
+    }
+
+#if NETSTANDARD1_0 || NETSTANDARD2_0
```
Not sure what this #if is for? This file should not be included for those defines.

peterfoot replied:
Sorry, that file is a placeholder copy-pasted. I'll add in the Email implementation today.

peterfoot replied:
Replaced with actual Win32 implementation

mattleibow commented on Xamarin.Essentials/Xamarin.Essentials.csproj:
```diff
@@ -106,6 +106,22 @@
+    </Reference>
+    <Reference Include="WindowsBase" />
+    <Reference Include="PresentationCore" />
+    <Reference Include="PresentationFramework" />
```
Just checking to see if we actually use the presentation bits for this project? We shouldn't be using any of the Xaml-related things. Also, this _should_ work for winforms without this?

peterfoot replied:
At the moment it uses WPF dependencies for MainThread, to support either WinForms or WPF will require some reworking but this is a sensible point. I will investigate alternative methods of identifying the thread.

DamianMehers replied:
Taking a dependency on Newtonsoft.Json in Essentials will make it unusable in some platforms such as WatchOS because the AOT generates a massive amount of code https://github.com/xamarin/xamarin-macios/issues/7250

Since only a limited number of types are supported in Preferences (bool, double, int, float, long, string, and DateTime) could you manually convert them to/from strings?

mattleibow commented on Xamarin.Essentials/Xamarin.Essentials.csproj:
```diff
@@ -1,6 +1,6 @@
 ﻿<Project Sdk="MSBuild.Sdk.Extras/2.0.54">
   <PropertyGroup>
-    <TargetFrameworks>netstandard1.0;netstandard2.0;Xamarin.iOS10;Xamarin.TVOS10;Xamarin.WatchOS10;MonoAndroid60;MonoAndroid70;MonoAndroid71;MonoAndroid80;MonoAndroid81;MonoAndroid90;MonoAndroid10.0;tizen40;</TargetFrameworks>
+    <TargetFrameworks>netstandard1.0;netstandard2.0;Xamarin.iOS10;Xamarin.TVOS10;Xamarin.WatchOS10;MonoAndroid60;MonoAndroid70;MonoAndroid71;MonoAndroid80;MonoAndroid81;MonoAndroid90;MonoAndroid10.0;tizen40;net461</TargetFrameworks>
```
We will need to think about how we do this... net461 works in too many places.

peterfoot replied:
I assume this will work with some more detailed tweaking to create multiple net461 dlls in the runtimes folder of the NuGet.

VladislavAntonyuk replied:
we should also add .Net Core WPF

DamianMehers commented on Xamarin.Essentials/Xamarin.Essentials.csproj:
```diff
@@ -106,6 +106,23 @@
+  <ItemGroup Condition=" $(TargetFramework.StartsWith('net461')) ">
+    <Compile Include="**\*.win32.*.cs" />
+    <Compile Include="**\*.win32.cs" />
+    <PackageReference Include="Newtonsoft.Json" Version="12.0.3" />
```
Taking a dependency on Newtonsoft.Json in Essentials will make it unusable in some platforms such as WatchOS because the AOT generates a massive amount of code xamarin/xamarin-macios#7250

Since only a limited number of types are supported in Preferences (bool, double, int, float, long, string, and DateTime) could you manually convert them to/from strings?

Mrnikbobjeff commented:
Geolocation and coding is available via the com interfaces in locationapi.h as far as I know.  Magnetometers are available in the sensorapi,h header via correspoding com objects
https://docs.microsoft.com/en-us/windows/win32/api/sensorsapi/

DamianMehers commented on Xamarin.Essentials/Preferences/Preferences.win32.cs:
```diff
@@ -1,43 +1,63 @@
     public static partial class Preferences
     {
         static readonly string settingsPath = Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.ApplicationData), AppInfo.Name, "settings.dat");
 
```
Any chance that AppInfo.Name will contain invalid chars for win32 file names, such as a ":" ?

peterfoot replied:
Good catch, the documentation for the attribute says "The assembly title is a friendly name, which can include spaces" but that's not particularly informative.

DamianMehers replied:
It could also be blank ... perhaps fallback to using the file name of the launching assembly if the sanitized name is blank?

``` c#
      var name = System.IO.Path.GetFileNameWithoutExtension(Assembly.GetEntryAssembly().CodeBase);
```

mattleibow commented on Xamarin.Essentials/Preferences/Preferences.win32.cs:
```diff
@@ -0,0 +1,133 @@
+    {
+        static readonly string settingsPath = Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.ApplicationData), AppInfo.Name, "settings.dat");
+
+        static readonly Dictionary<string, Dictionary<string, object>> preferences = new Dictionary<string, Dictionary<string, object>>();
```
This is a tad bit nit-picky, but it might be nicer to just add a usings to the top of the file:

```csharp
using PreferencesDictionary = System.Collections.Generic.Dictionary<string, System.Collections.Generic.Dictionary<string, object>>;
```
This way, we don't have angle brackets all over the show.

mattleibow replied:
This is a tad bit nit-picky, but it might be nicer to just add a usings to the top of the file:

```csharp
using PreferencesDictionary = System.Collections.Generic.Dictionary<string, System.Collections.Generic.Dictionary<string, object>>;
```
This way, we don't have angle brackets all over the show.

mattleibow commented on Xamarin.Essentials/Preferences/Preferences.win32.cs:
```diff
@@ -0,0 +1,133 @@
+
+        static bool PlatformContainsKey(string key, string sharedName)
+        {
+            if (preferences.ContainsKey(CleanSharedName(sharedName)))
```
This could be a TryGetValue:

```csharp
return preferences.TryGetValue(sharedName, out var inner) && inner.ContainsKey(key);
```

mattleibow commented on Xamarin.Essentials/Preferences/Preferences.win32.cs:
```diff
@@ -0,0 +1,133 @@
+            {
+                var formatter = new BinaryFormatter();
+                formatter.Serialize(stream, preferences);
+                stream.Close();
```
The dispose will close here.

mattleibow commented on Xamarin.Essentials/Preferences/Preferences.win32.cs:
```diff
@@ -0,0 +1,133 @@
+                    {
+                        var formatter = new BinaryFormatter();
+                        var readPreferences = (Dictionary<string, Dictionary<string, object>>)formatter.Deserialize(stream);
+                        stream.Close();
```
no need to close, the dispose will do it

mattleibow commented on Xamarin.Essentials/Preferences/Preferences.win32.cs:
```diff
@@ -0,0 +1,133 @@
+
+        static void PlatformRemove(string key, string sharedName)
+        {
+            if (preferences.ContainsKey(CleanSharedName(sharedName)))
```
Maybe use the TryGetValue here so save a lookup.

```csharp
if (prefs.TryGetValue(sN, out var inner)) {
    inner.Remove(key);
    Save();
}
```

mattleibow commented on Xamarin.Essentials/Preferences/Preferences.win32.cs:
```diff
@@ -0,0 +1,133 @@
+                preferences.Add(CleanSharedName(sharedName), new Dictionary<string, object>());
+            }
+
+            if (preferences[CleanSharedName(sharedName)].ContainsKey(key))
```
Maybe use a TryGetValue for these.

In fact, just use the `[key] = value` here as it will add any missing keys. No need to check and then branch code.

mattleibow commented on Xamarin.Essentials/Preferences/Preferences.win32.cs:
```diff
@@ -0,0 +1,133 @@
+            {
+                if (preferences[CleanSharedName(sharedName)].ContainsKey(key))
+                {
+                    return (T)preferences[CleanSharedName(sharedName)][key];
```
Maybe a `if (prefs.TryGetValue() && inner.TryGetValue())`

Mrnikbobjeff commented on Xamarin.Essentials/Preferences/Preferences.win32.cs:
```diff
@@ -0,0 +1,133 @@
+                        var readPreferences = (Dictionary<string, Dictionary<string, object>>)formatter.Deserialize(stream);
+                        stream.Close();
+
+                        if (readPreferences is object)
```
Just check for null the normal way. https://github.com/xamarin/Essentials/blob/master/CONTRIBUTING.md

Mrnikbobjeff commented on Xamarin.Essentials/Preferences/Preferences.win32.cs:
```diff
@@ -1,16 +1,18 @@
+using System.Reflection;
 using System.Runtime.Serialization;
 using System.Runtime.Serialization.Formatters.Binary;
+using PreferencesDictionary = System.Collections.Generic.Dictionary<string, System.Collections.Generic.Dictionary<string, object>>;
```
Something to consider would be using IStrongBox<T> as value container, as we can ellide boxing/unboxing some of the time this way: IStrongBox<T> {public T Value;}
When reading a key we can instead do (value as IStrongBox<T>).Value. This pattern is common in storing value types without boxing as often.

Mrnikbobjeff commented:
An easy grab is Telephone and sms, simply invoke the launcher with the corresponding url. Same goes for sms. An improvement to launcher can be made by checking the registry for a corrsponding HKEY_CLASSES_ROOT\%protocol%\shell\open\command registry key to check whether launching said url is supported

Mrnikbobjeff commented on Xamarin.Essentials/Geolocation/Geolocation.win32.cs:
```diff
@@ -0,0 +1,52 @@
+
+namespace Xamarin.Essentials
+{
+    public static partial class Geolocation
```
Consider using the com interfaces for this, we avoid having a dependency for this. The actual implementation uses the COM service as well. Very Optional though as the C# wrapper is quite a lot easier to deal with.

peterfoot replied:
@Mrnikbobjeff The System.Devices dll is a built in part of the .NET framework so doesn't introduce a dependency, using the COM interfaces will require more code to achieve the same result.

Redth commented:
A few notes as @mattleibow and I have been discussing:

1. Preferences should really serialize to json or some text format, as binary serialization has some dragons especially moving to .net 5
2. SecureStorage - we might be able to do something on a per user basis here, Potentially using [CNG-DPAPI](https://docs.microsoft.com/en-us/windows/win32/seccng/cng-dpapi)
3. We should consider how to handle Local vs Roaming profiles for preferences, and Secure Storage,  etc.

peterfoot commented:
@Redth Thanks, Json was rejected before because Json.net adds a dependency and System.Text.Json not available on older .NET versions. The only consistently available text format in-box is XML...

DPAPI/ProtectedData for SecureStorage makes sense, again physical storage is an issue.

I assume all of these will be non-roaming as the other platforms don't roam their preferences?

Redth commented:
What about DataContractJsonSerializer? Or yeah System.Text.Json. 

Other platforms implicitly can ‘roam’ their data depending on where it’s stored

peterfoot commented:
@Redth - yes DataContractJsonSerializer looks like a good option.

Looking at the current implementations for Essentials they are all using private/local storage. I assume for WPF we want to match the behaviour of the existing UWP implementation where possible.

mattleibow commented:
System.Text.Json does work on netstandard2.0 and net4.6.1

https://www.nuget.org/packages/System.Text.Json

peterfoot commented:
@mattleibow Given the simple use-case here is it worth using System.Text.Json (introducing a NuGet dependency) or stick with the built-in DataContractJsonFormatter?

mattleibow commented:
@peterfoot I would just use the contract serializer for now. Just wanted to say that it should work... But as I type this, I somehow feel that there was some issue with the serializer... Like dates... Not sure if anyone knows... Maybe it was around deserialization. But since we serialize and deserialize ourselves, we should be good.

Mrnikbobjeff commented:
@mattleibow as we specify which types we accept (and all types we support are sealed) we do not encounter most serialisation issues such as ignoring validation when deserializing objects or expecting different types upon deserialization. Any text format should work without problems.

Mrnikbobjeff commented on Xamarin.Essentials/Battery/Battery.win32.cs:
```diff
@@ -0,0 +1,128 @@
+
+        static class NativeMethods
+        {
+            [DllImport("Kernel32")]
```
Missing exactspelling attribute for efficient stub lookup

Mrnikbobjeff commented on Xamarin.Essentials/DeviceDisplay/DeviceDisplay.win32.cs:
```diff
@@ -0,0 +1,124 @@
+
+        static class NativeMethods
+        {
+            [DllImport("Kernel32", SetLastError = true)]
```
Add ExactSpelling

Mrnikbobjeff commented on Xamarin.Essentials/DeviceDisplay/DeviceDisplay.win32.cs:
```diff
@@ -0,0 +1,124 @@
+                DISPLAY_REQUIRED = 0x00000002,
+            }
+
+            [DllImport("user32")]
```
Add ExactSpelling

Mrnikbobjeff commented on Xamarin.Essentials/DeviceDisplay/DeviceDisplay.win32.cs:
```diff
@@ -0,0 +1,124 @@
+            [DllImport("user32")]
+            internal static extern IntPtr GetDesktopWindow();
+
+            [DllImport("user32")]
```
Add ExactSpelling

Mrnikbobjeff commented on Xamarin.Essentials/DeviceDisplay/DeviceDisplay.win32.cs:
```diff
@@ -0,0 +1,124 @@
+                NEAREST = 2,
+            }
+
+            [DllImport("user32")]
```
Add ExactSpelling, switch to GetMonitorInfoW for less string marshalling overhead

Mrnikbobjeff commented on Xamarin.Essentials/DeviceDisplay/DeviceDisplay.win32.cs:
```diff
@@ -0,0 +1,124 @@
+            }
+
+            [DllImport("gdi32")]
+            internal static extern IntPtr CreateDC(string lpszDriver, string lpszDevice, string lpszOutput, IntPtr lpInitData);
```
Switch to CreateDCW

Mrnikbobjeff commented on Xamarin.Essentials/DeviceDisplay/DeviceDisplay.win32.cs:
```diff
@@ -0,0 +1,124 @@
+            [DllImport("gdi32")]
+            internal static extern IntPtr CreateDC(string lpszDriver, string lpszDevice, string lpszOutput, IntPtr lpInitData);
+
+            [DllImport("gdi32")]
```
ExactSpelling

Mrnikbobjeff commented on Xamarin.Essentials/DeviceDisplay/DeviceDisplay.win32.cs:
```diff
@@ -0,0 +1,124 @@
+            internal static extern bool DeleteDC(IntPtr hdc);
+
+            [DllImport("gdi32")]
+            internal static extern int GetDeviceCaps(IntPtr hdc, DeviceCap nIndex);
```
ExactSpelling

Mrnikbobjeff reviewed (commented):
Some interop droplets

peterfoot commented:
Thanks @Mrnikbobjeff I think I've fixed up all these interop points. I've also reworked it against the latest in main. And added the FilePicker implementation for Win32 😀

Mrnikbobjeff commented:
Thank you for taking time to implement these changes. I would provide a Screenshot implementation of you have not started working on it yet, that is a very easy task. I will then create a merge request into this branch

peterfoot commented:
@Mrnikbobjeff that would be great! It's a feature which wasn't in Essentials when I first created the PR and will be really useful.

Mrnikbobjeff commented:
> @Mrnikbobjeff that would be great! It's a feature which wasn't in Essentials when I first created the PR and will be really useful.

You have your PR, I also added a missing charset.unicode parameter I missed in the review

peterfoot commented:
@Mrnikbobjeff awesome thanks!

Mrnikbobjeff commented:
@peterfoot I created a second pr to check for sms and telephone capabilities.

Mrnikbobjeff commented:
I finished working on the first sensor implementation, I believe we are able to extend support to the following sensors:
Accelerometer, Barometer, Compass, Gyroscope, Magnetometer, OrientationSensor, Vibration (when a xinput device is present). They all offer com interfaces which sensor device developers can target

Mrnikbobjeff commented:
I polished the com interop a little and startet a pr. You can see the sensor groundwork [here](https://github.com/inthehand/Essentials/blob/5694bfa6a761db187141b2261ff3676483e6be8b/Xamarin.Essentials/Barometer/Barometer.win32.cs). Adding the other sensors should be easy now, but it would be wonderful if someone checks out the path we are taking here as sensor support is approximately 3k loc

peterfoot commented:
@Mrnikbobjeff I suppose the thing is - how many Windows devices will have those sensors? And of those, how many running version prior to Windows 10? I'm looking at NuGet runtime support so we can generate multiple .NET Framework assemblies - this will ensure there is a separate windows and linux version. Also means there can be a Windows 7 and above version and a Windows 10 specific version which can use the UWP APIs - I think all the sensor APIs don't require a package id so can be used from desktop apps.

Mrnikbobjeff commented:
> @Mrnikbobjeff I suppose the thing is - how many Windows devices will have those sensors? And of those, how many running version prior to Windows 10? I'm looking at NuGet runtime support so we can generate multiple .NET Framework assemblies - this will ensure there is a separate windows and linux version. Also means there can be a Windows 7 and above version and a Windows 10 specific version which can use the UWP APIs - I think all the sensor APIs don't require a package id so can be used from desktop apps.

The sensor API is available from Windows 7 onward

peterfoot commented:
@Mrnikbobjeff yes, but my question was that if we can use the already written UWP API for Windows 10, is it worth writing the Win32 ones. It would really be only a subset of tablets/laptops with the hardware and how many would be running Windows 7/8?
The macOS implementation doesn't have sensors as I don't think they are as relevant for desktop apps.

Mrnikbobjeff commented:
> The macOS implementation doesn't have sensors as I don't think they are as relevant for desktop apps.

MacOs does not offer an abstraction over general p&p devices which device drivers can target for these sensors. That is the limiting factor IMHO. Also, writing new sensors once the base implementation is done is achieved in [less than 20 loc](https://github.com/inthehand/Essentials/blob/5694bfa6a761db187141b2261ff3676483e6be8b/Xamarin.Essentials/Barometer/Barometer.win32.cs#L2838) up to and including BarometerSensor

peterfoot commented:
@Mrnikbobjeff I'm not against implementing sensor APIs, I just wonder how many devices will be able to take advantage of them

Mrnikbobjeff commented:
> @Mrnikbobjeff I'm not against implementing sensor APIs, I just wonder how many devices will be able to take advantage of them

Very few at best, the most common sensors appear to be biometrics and luminosity sensors. My work device seems to have them both, and biometrics seem to be supported by other devices (primarily laptops) as well.

follesoe commented:
Just want to say this would be highly appreciated and valued. Given that WPF is part of Xamarin.Forms (even if only as preview) it makes perfect sense to get Essentials supported on WPF as well. At Blueye we are using WPF over UWP due to limitations in some third-party libraries which prevents us from using Store Distribution, and side-loading of UWP is not as nice/trivial as with WPF.

I would be glad to help out with the implementation of some of the missing pieces. Would be great with some guidance on what would be required in order to have this merged (i.e. does all APIs need to be ported)?

Thank you for your effort so far @peterfoot! It has already been of value as we have adopted parts of your implementation already while awaiting the merge of this PR.

arivera12 commented:
Oh yes! Xamarin essentials is coming to WPF! When this is planned to be released?

acffordyce973 commented:
Are there any updates for Xamarin Essentials supporting WPF?

peterfoot commented:
@acffordyce973 the PR was a bit stale so I've updated against the latest main branch so it should unblock the conflicts at least.
